src/Control/Monad/HashCons.hs

```diff
@@ -92,6 +94,20 @@ hashCons x = HashConsT $ do
           newRef <$ modify (Bimap.insert newRef x)
     Just ref -> pure ref
 
+-- | Given a value of type @r@, fetch the cached @e@ value, if it exists.
+--
+-- @since 1.0.0
+lookupRef ::
+  forall (r :: Type) (e :: Type) (m :: Type -> Type).
+  (Monad m, Ord e, Ord r) =>
+  r ->
+  HashConsT r e m e
+lookupRef r =
+  HashConsT
+    ( -- This shouldn't fail, as long as users of HashConsT don't construct @r@ out of thin air
+      fromJust . Bimap.lookup r <$> get
```

Comment:
@kozross do you think this is reasonable?

Reply:
This is not safe in general. There is no guarantee arbitrary `r` values are mapped to anything in the state. In our _specific_ case, we can be sure that this works because `Id` construction is tightly controlled. It's probably better to write an `Id`-specific helper, since all this is really doing is `fromJust`ing.